File: iosApp/iosApp/ui/offer/PreviewableOfferTruthSource.swift
//
//  PreviewableOfferTruthSource.swift
//  iosApp
//
//  Created by jimmyt on 7/12/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import BigInt
import Foundation
import web3swift

/**
 A `UIOfferTruthSource` implementation used for previewing user interfaces.
 */
class PreviewableOfferTruthSource: UIOfferTruthSource {
    
    /**
     Initializes a new `PreviewableOfferTruthSource` with sample offers.
     */
    init() {
        offers = Offer.sampleOffers
    }
    
    /**
     A dictionary mapping offer IDs (as `UUID`s) to `Offer`s, which is the single source of truth for all open-offer-related data.
     */
    @Published var offers: [UUID : Offer]
    
    /**
     The current service fee rate, or `nil` if the current service fee rate is not known.
     */
    @Published var serviceFeeRate: BigUInt?
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    var isGettingServiceFeeRate: Bool = false
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func updateServiceFeeRate() {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    var approvingTransferToOpenOfferState: TokenTransferApprovalState = .none
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    var approvingTransferToOpenOfferError: Error?
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func createApproveTokenTransferToOpenOfferTransaction(chainID: BigUInt, stablecoin: EthereumAddress?, stablecoinInformation: StablecoinInformation?, minimumAmount: Decimal, maximumAmount: Decimal, securityDepositAmount: Decimal, direction: OfferDirection, settlementMethods: [SettlementMethod], createdTransactionHandler: @escaping (EthereumTransaction) -> Void, errorHandler: @escaping (Error) -> Void) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func approveTokenTransferToOpenOffer(chainID: BigUInt, stablecoin: EthereumAddress?, stablecoinInformation: StablecoinInformation?, minimumAmount: Decimal, maximumAmount: Decimal, securityDepositAmount: Decimal, direction: OfferDirection, settlementMethods: [SettlementMethod], approveTokenTransferToOpenOfferTransaction: EthereumTransaction?) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func createOpenOfferTransaction(offer: Offer, createdTransactionHandler: @escaping (EthereumTransaction) -> Void, errorHandler: @escaping (Error) -> Void) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func openOffer(offer: Offer, offerOpeningTransaction: EthereumTransaction?) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func createOffer(chainID: BigUInt, stablecoin: EthereumAddress?, stablecoinInformation: StablecoinInformation?, minimumAmount: Decimal, maximumAmount: Decimal, securityDepositAmount: Decimal, direction: OfferDirection, settlementMethods: [SettlementMethod]) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func openOffer(chainID: BigUInt, stablecoin: EthereumAddress?, stablecoinInformation: StablecoinInformation?, minimumAmount: Decimal, maximumAmount: Decimal, securityDepositAmount: Decimal, direction: OfferDirection, settlementMethods: [SettlementMethod]) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func cancelOffer(_ offer: Offer) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func createCancelOfferTransaction(offer: Offer, createdTransactionHandler: (EthereumTransaction) -> Void, errorHandler: (Error) -> Void) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func cancelOffer(offer: Offer, offerCancellationTransaction: EthereumTransaction?) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func createEditOfferTransaction(offer: Offer, newSettlementMethods: [SettlementMethod], createdTransactionHandler: @escaping (EthereumTransaction) -> Void, errorHandler: @escaping (Error) -> Void) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func editOffer(offer: Offer, newSettlementMethods: [SettlementMethod], offerEditingTransaction: EthereumTransaction?) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func editOffer(offer: Offer, newSettlementMethods: [SettlementMethod]) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func createApproveTokenTransferToTakeOfferTransaction(
        offer: Offer,
        takenSwapAmount: Decimal,
        makerSettlementMethod: SettlementMethod?,
        takerSettlementMethod: SettlementMethod?,
        createdTransactionHandler: @escaping (EthereumTransaction) -> Void,
        errorHandler: @escaping (Error) -> Void
    ) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func approveTokenTransferToTakeOffer(
        offer: Offer,
        takenSwapAmount: Decimal,
        makerSettlementMethod: SettlementMethod?,
        takerSettlementMethod: SettlementMethod?,
        approveTokenTransferToTakeOfferTransaction: EthereumTransaction?
    ) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func createTakeOfferTransaction(
        offer: Offer,
        takenSwapAmount: Decimal,
        makerSettlementMethod: SettlementMethod?,
        takerSettlementMethod: SettlementMethod?,
        createdTransactionAndKeyPairHandler: @escaping (EthereumTransaction, KeyPair) -> Void,
        errorHandler: @escaping (Error) -> Void
    ) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func takeOffer(
        offer: Offer,
        takenSwapAmount: Decimal,
        makerSettlementMethod: SettlementMethod?,
        takerSettlementMethod: SettlementMethod?,
        keyPair: KeyPair?,
        offerTakingTransaction: EthereumTransaction?
    ) {}
    
    /**
     Not used since this class is for previewing user interfaces, but required for adoption of `UIOfferTruthSource`.
     */
    func takeOffer(offer: Offer, takenSwapAmount: Decimal, makerSettlementMethod: SettlementMethod?, takerSettlementMethod: SettlementMethod?) {}
    
}

File: androidApp/src/main/java/com/commuto/interfacemobile/android/ui/offer/PreviewableOfferTruthSource.kt
package com.commuto.interfacemobile.android.ui.offer

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.snapshots.SnapshotStateMap
import com.commuto.interfacemobile.android.key.keys.KeyPair
import com.commuto.interfacemobile.android.offer.Offer
import com.commuto.interfacemobile.android.offer.OfferDirection
import com.commuto.interfacemobile.android.offer.TokenTransferApprovalState
import com.commuto.interfacemobile.android.settlement.SettlementMethod
import com.commuto.interfacemobile.android.ui.StablecoinInformation
import org.web3j.crypto.RawTransaction
import java.math.BigDecimal
import java.math.BigInteger
import java.util.*

/**
 * A [UIOfferTruthSource] implementation used for previewing user interfaces.
 *
 * @property offers A [SnapshotStateMap] mapping [UUID]s to [Offer]s, which acts as a single source of truth for all
 * offer-related data.
 * @property serviceFeeRate The current
 * [service fee rate](https://github.com/jimmyneutront/commuto-whitepaper/blob/main/commuto-whitepaper.txt) as a
 * percentage times 100, or `null` if the current service fee rate is not known.
 * @property isGettingServiceFeeRate Indicates whether this is currently getting the current service fee rate. This will
 * always be false, this class is used only for previewing user interfaces.
 */
class PreviewableOfferTruthSource: UIOfferTruthSource {
    override var offers = SnapshotStateMap<UUID, Offer>().also { map ->
        Offer.sampleOffers.map {
            map[it.id] = it
        }
    }
    override var serviceFeeRate: MutableState<BigInteger?> = mutableStateOf(null)
    override var isGettingServiceFeeRate = mutableStateOf(false)

    /**
     * Not used since this class is for previewing user interfaces, but required for implementing [UIOfferTruthSource].
     */
    override val approvingTransferToOpenOfferState = mutableStateOf(TokenTransferApprovalState.NONE)

    /**
     * Not used since this class is for previewing user interfaces, but required for implementing [UIOfferTruthSource].
     */
    override var approvingTransferToOpenOfferException: Exception? = null

    /**
     * Does nothing, since this class is only used for previewing user interfaces.
     */
    override fun addOffer(offer: Offer) {}

    /**
     * Does nothing, since this class is only used for previewing user interfaces.
     */
    override fun removeOffer(id: UUID) {}

    /**
     * Does nothing, since this class is only used for previewing user interfaces.
     */
    override fun updateServiceFeeRate() {}

    /**
     * Not used since this class is for previewing user interfaces, but required for implementing [UIOfferTruthSource].
     */
    override fun createApproveToOpenTransaction(
        stablecoin: String?,
        stablecoinInformation: StablecoinInformation?,
        minimumAmount: BigDecimal,
        maximumAmount: BigDecimal,
        securityDepositAmount: BigDecimal,
        direction: OfferDirection?,
        settlementMethods: List<SettlementMethod>,
        createdTransactionHandler: (RawTransaction) -> Unit,
        exceptionHandler: (Exception) -> Unit,
    ) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for implementing [UIOfferTruthSource].
     */
    override fun approveTokenTransferToOpenOffer(
        chainID: BigInteger,
        stablecoin: String?,
        stablecoinInformation: StablecoinInformation?,
        minimumAmount: BigDecimal,
        maximumAmount: BigDecimal,
        securityDepositAmount: BigDecimal,
        direction: OfferDirection?,
        settlementMethods: List<SettlementMethod>,
        approveTokenTransferToOpenOfferTransaction: RawTransaction?,
    ) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for implementing [UIOfferTruthSource].
     */
    override fun createOpenOfferTransaction(
        offer: Offer,
        createdTransactionHandler: (RawTransaction) -> Unit,
        exceptionHandler: (Exception) -> Unit
    ) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for implementing [UIOfferTruthSource].
     */
    override fun openOffer(
        offer: Offer,
        offerOpeningTransaction: RawTransaction
    ) {}

    /**
     * Does nothing since this class is only used for previewing user interfaces, but is required for implementing
     * [UIOfferTruthSource]
     */
    override fun openOffer(
        chainID: BigInteger,
        stablecoin: String?,
        stablecoinInformation: StablecoinInformation?,
        minimumAmount: BigDecimal,
        maximumAmount: BigDecimal,
        securityDepositAmount: BigDecimal,
        direction: OfferDirection?,
        settlementMethods: List<SettlementMethod>
    ) {}

    /**
     * Does nothing since this class is only used for previewing user interfaces, but is required for implementing
     * [UIOfferTruthSource]
     */
    @Deprecated("Use the new offer pipeline with improved transaction state management")
    override fun cancelOffer(offer: Offer) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for adoption of [UIOfferTruthSource].
     */
    override fun createCancelOfferTransaction(
        offer: Offer,
        createdTransactionHandler: (RawTransaction) -> Unit,
        exceptionHandler: (Exception) -> Unit
    ) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for adoption of [UIOfferTruthSource].
     */
    override fun cancelOffer(offer: Offer, offerCancellationTransaction: RawTransaction?) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for adoption of [UIOfferTruthSource].
     */
    override fun createEditOfferTransaction(
        offer: Offer,
        newSettlementMethods: List<SettlementMethod>,
        createdTransactionHandler: (RawTransaction) -> Unit,
        exceptionHandler: (Exception) -> Unit
    ) {}

    /**
     * Not used since this class is for previewing user interfaces, but required for adoption of [UIOfferTruthSource].
     */
    override fun editOffer(
        offer: Offer,
        newSettlementMethods: List<SettlementMethod>,
        offerEditingTransaction: RawTransaction?
    ) {}

    /**
     * Does nothing since this class is only used for previewing user interfaces, but is required for implementing
     * [UIOfferTruthSource]
     */
    override fun editOffer(
        offer: Offer,
        newSettlementMethods: List<SettlementMethod>
    ) {}

    /**
     * Does nothing since this class is only used for previewing user interfaces, but is required for implementing
     * [UIOfferTruthSource]
     */
    override fun createApproveTokenTransferToTakeOfferTransaction(
        offer: Offer,
        takenSwapAmount: BigDecimal,
        makerSettlementMethod: SettlementMethod?,
        takerSettlementMethod: SettlementMethod?,
        createdTransactionHandler: (RawTransaction) -> Unit,
        exceptionHandler: (Exception) -> Unit
    ) {}

    /**
     * Does nothing since this class is only used for previewing user interfaces, but is required for implementing
     * [UIOfferTruthSource]
     */
    override fun approveTokenTransferToTakeOffer(
        offer: Offer,
        takenSwapAmount: BigDecimal,
        makerSettlementMethod: SettlementMethod?,
        takerSettlementMethod: SettlementMethod?,
        approveTokenTransferToOpenOfferTransaction: RawTransaction?
    ) {}

    /**
     * Does nothing since this class is only used for previewing user interfaces, but is required for implementing
     * [UIOfferTruthSource]
     */
    override fun createTakeOfferTransaction(
        offer: Offer,
        takenSwapAmount: BigDecimal,
        makerSettlementMethod: SettlementMethod?,
        takerSettlementMethod: SettlementMethod?,
        createdTransactionAndKeyPairHandler: (RawTransaction, KeyPair) -> Unit,
        exceptionHandler: (Exception) -> Unit
    ) {}

    /**
     * Does nothing since this class is only used for previewing user interfaces, but is required for implementing
     * [UIOfferTruthSource]
     */
    override fun takeOffer(
        offer: Offer,
        takenSwapAmount: BigDecimal,
        makerSettlementMethod: SettlementMethod?,
        takerSettlementMethod: SettlementMethod?,
        keyPair: KeyPair?,
        offerTakingTransaction: RawTransaction?,
    ) {}

    /**
     * Does nothing since this class is only used for previewing user interfaces, but is required for implementing
     * [UIOfferTruthSource]
     */
    override fun takeOffer(
        offer: Offer,
        takenSwapAmount: BigDecimal,
        makerSettlementMethod: SettlementMethod?,
        takerSettlementMethod: SettlementMethod?,
    ) {}
}